### ISSUE TITLE
Redesign commands

### ISSUE THREAD

**Issue by SoniEx2** [NOVA-Team/NOVA-Core#224]
Current idea:

```
List<Arg> ->
  get(int index) ->
    has();
    as<primitive type>(); (e.g. asInt(); asLong(); etc)
    as(Class<? extends ArgConverter<T>> clazz);
```

Questions:

1. Should we have `asString`?
2. Should `as` take any `Class`, and if there's no converter for it then cause a runtime error? (This seems like bad API design, but it'd let you pass in `String.class`)

**Comment by RX14** [comment]
What's the use of the `has()` function?

**Comment by SoniEx2** [comment]
Instead of returning a proper List, we'd return a wrapped List that returns an empty Arg for args out of bounds.

**Comment by RX14** [comment]
That appears to be a really bad idea to me.

**Comment by SoniEx2** [comment]
@RX14 Could always use Map<Integer, Arg> instead.

**Comment by RX14** [comment]
Could always expect programmers to be able to use `List#size`...

**Comment by SoniEx2** [comment]
@RX14 Returning a valid value on out of bounds without changing List.size should be safe. Not sure if it's allowed by the List.get contract but w/e, it doesn't matter, your code shouldn't rely on OutOfBoundsExceptions anyway.

**Comment by RX14** [comment]
No, this clearly breaks the contract of `List<T>`. Nobody expects `List` to do that. If you want to create an interface like this, then we should create an `Args` class instead of `List<Arg>`.

**Comment by SoniEx2** [comment]
@RX14 This isn't a question of "does it break the contract?" this is a question of "does it work?". And it very well does.

**Comment by SoniEx2** [comment]
With `Args.asList()`?

**Comment by RX14** [comment]
Maybe, but not as the main API, it should have `size()` and `get(int)` (like `List<T>`), but return `EmptyArg` classes instead of index error.

**Comment by SoniEx2** [comment]
Should `Args.asList()` return a `List<Arg>` *view* of the args? Or should it return a `List<Arg>` *copy* of the args?

**Comment by RX14** [comment]
@SoniEx2 semantics, return a Collections.immutableList view of it.

**Comment by SoniEx2** [comment]
@RX14 *why?*

**Comment by RX14** [comment]
Why not?

**Comment by SoniEx2** [comment]
@RX14 Well idk but why should `Args` be immutable?

**Comment by RX14** [comment]
Because they are in reality immutable, you run the command with arguments, and you can't retroactively change what the user meant.

**Comment by SoniEx2** [comment]
@RX14 Technically, you're given an `Args`. The `Args` is never used anywhere else so you're free to do with it as you like.

**Comment by RX14** [comment]
Technically, you're wrong.

**Comment by SoniEx2** [comment]
@RX14 Why limit the user?

**Comment by RX14** [comment]
because it's the true state of the world that you cannot retroactively change the arguments the user passed into a command. That's time travel and pretending otherwise has no benefits and doesn't show the true state of the user interaction.

**Comment by SoniEx2** [comment]
@RX14 Everything is mutable if you put enough hacks into it. And you might want to pass the `Args` forward after some tweaking.

**Comment by RX14** [comment]
@SoniEx2 I have explaied why multiple times, and I'm pretty sure the other developers will agree with me that having Args be mutable "feels wrong". You can construct your own Args objects with your own list.

**Comment by calclavia** [comment]
I agree with having Args be immutable.

**Comment by ExE-Boss** [comment]
Should there be a `CommandFactory` or should it just be a `Command` instance?

**Comment by SoniEx2** [comment]
I'm not sure the purpose of a CommandFactory other than to waste my RAM?
